### DISCUSSION TITLE
Does gatsby-source-contentful support Gatsby 4?

### DISCUSSION THREAD

**Discussion by cameronmcefee** [gatsbyjs/gatsby#33933] [Help] [answered]
What version of gatsby-source-contentful should I be using with Gatsby 4?

After updating to Gatsby 4 and the latest version of gatsby-source-contentful (6.1.2) I'm seeing errors about image files not being found in the cache.

<details>
<summary>See error</summary>

```
/Users/cameronmcefee/github/static-sites/node_modules/yoga-layout-prebuilt/yoga-layout/build/Release/nbind.js:53
        throw ex;
        ^

Error: ENOENT: no such file or directory, open '/Users/cameronmcefee/github/static-sites/packages/blog/.cache/caches/gatsby-source-contentful/749975118e1502d8238e5e8d4293b796-3/ecolaunch-hero-static.png'
Emitted 'error' event on ReadStream instance at:
    at /Users/cameronmcefee/github/static-sites/node_modules/graceful-fs/graceful-fs.js:292:14
    at /Users/cameronmcefee/github/static-sites/node_modules/graceful-fs/graceful-fs.js:343:16
    at FSReqCallback.oncomplete (fs.js:180:23) {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/Users/cameronmcefee/github/static-sites/packages/blog/.cache/caches/gatsby-source-contentful/749975118e1502d8238e5e8d4293b796-3/ecolaunch-hero-static.png' ERROR #85928 

An error occurred during parallel query running.
Go here for troubleshooting tips: https://gatsby.dev/pqr-feedback
```

</details>

The [linked discussion](https://gatsby.dev/pqr-feedback) in the error message suggests I should be using `gatsby-source-contentful@alpha-pqr-compat` which currently resolves to `gatsby-source-contentful@5.10.0-alpha-pqr-compat.2` which is a major version behind.

Doing so, I run into different errors related to VipsJpeg having issues with the files it's receiving from Contentful.

<details>
<summary>See error</summary>

```
There was an error in your GraphQL query:

VipsJpeg: Premature end of input file
VipsJpeg: Premature end of input file
VipsJpeg: out of order read at line 128
VipsJpeg: out of order read at line 128
VipsJpeg: Premature end of input file
VipsJpeg: out of order read at line 128

  74 |   }
  75 | }
  76 |
  77 | fragment BlogCategory on ContentfulBlogCategory {
  78 |   title
  79 |   slug
  80 |   metaImage {
  81 |     ...Image
  82 |   }
  83 |   heroImage {
> 84 |     gatsbyImageData
     |     ^
  85 |   }
  86 |   body {
  87 |     childMarkdownRemark {
  88 |       html
  89 |     }
  90 |   }
  91 | }
  92 |
  93 | fragment EnhancedImage on ContentfulEnhancedImage {
  94 |   id
```
</details>

In either case, I know the issue is mine to resolve, but I'm not sure which version is the one to use and which problem I need to be troubleshooting.

I unfortunately can't provide repro example but since I'm not asking for troubleshooting help, hopefully that won't get me disqualified 😬 .

Thanks for your help.

**Comment by LekoArts** [comment, marked as answer]
Hi!

Yes, it supports it and you should use v6 of contentful. The error you're seeing is a bug that will (hopefully) be fixed with https://github.com/gatsbyjs/gatsby/pull/33932

**Reply by cameronmcefee** [reply]
Ah thanks @LekoArts, it's nice to hear the problem isn't me for a change 😄.

**Reply by LekoArts** [reply]
Please update all dependencies and try again :) We released the PR I mentioned

**Reply by cameronmcefee** [reply]
Regretfully, this doesn't resolve things for me. Using gatsby@4.1.4 and gatsby-source-contentful@6.1.3 and a run of `gatsby clean` I get:

```
/Users/cameronmcefee/github/static-sites/node_modules/yoga-layout-prebuilt/yoga-layout/build/Release/nbind.js:53
        throw ex;
        ^

Error: ENOENT: no such file or directory, open '/Users/cameronmcefee/github/static-sites/packages/blog/.cache/caches/gatsby-source-contentful/572b8189cd9d06d6f4d9e09a5bae6a1c-1/performance3-hero.jpg'
Emitted 'error' event on ReadStream instance at:
    at /Users/cameronmcefee/github/static-sites/node_modules/graceful-fs/graceful-fs.js:292:14
    at /Users/cameronmcefee/github/static-sites/node_modules/graceful-fs/graceful-fs.js:343:16
    at FSReqCallback.oncomplete (fs.js:180:23) {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/Users/cameronmcefee/github/static-sites/packages/blog/.cache/caches/gatsby-source-contentful/572b8189cd9d06d6f4d9e09a5bae6a1c-1/performance3-hero.jpg'

 ERROR #85928 

An error occurred during parallel query running.
Go here for troubleshooting tips: https://gatsby.dev/pqr-feedback

...
```

Looking at my .cache, the error says it's looking for:

`572b8189cd9d06d6f4d9e09a5bae6a1c-1/performance3-hero.jpg`

but the folder on disk is:

`572b8189cd9d06d6f4d9e09a5bae6a1c-3/performance3-hero.jpg`

I don't know if the fact that the hash is the same but the last int is different is meaningful, but FYI.

Is there a better place for this? Should I create an issue?

**Reply by TylerBarnes** [reply]
Thanks for the report @cameronmcefee. We thought we had fixed the issue this morning but it seems there were perhaps multiple related issues with this code. Is your site open source or is there a way to share it with us so we can reproduce the issue?

**Reply by cameronmcefee** [reply]
The site is unfortunately not open. I might be able to grant limited contributor access (would need to check) but wouldn't be able to share the Contentful credentials required to fetch the content that's causing the errors. Would that be helpful or is it moot without the Contentful content?

**Reply by wardpeet** [reply]
https://github.com/gatsbyjs/gatsby/pull/33975 should fix the issue

**Reply by wardpeet** [reply]
Published in latest v4
```
Successfully published:
 - babel-plugin-remove-graphql-queries@4.1.3
 - babel-preset-gatsby@2.1.3
 - gatsby-cli@4.1.4
 - gatsby-core-utils@3.1.3
 - gatsby-page-utils@2.1.3
 - gatsby-plugin-benchmark-reporting@2.1.3
 - gatsby-plugin-gatsby-cloud@4.1.3
 - gatsby-plugin-image@2.1.3
 - gatsby-plugin-manifest@4.1.4
 - gatsby-plugin-mdx@3.1.4
 - gatsby-plugin-offline@5.1.4
 - gatsby-plugin-page-creator@4.1.4
 - gatsby-plugin-preload-fonts@3.1.3
 - gatsby-plugin-sharp@4.1.4
 - gatsby-plugin-typescript@4.1.3
 - gatsby-recipes@1.1.3
 - gatsby-remark-images@6.1.4
 - gatsby-source-contentful@6.1.4
 - gatsby-source-drupal@5.1.4
 - gatsby-source-filesystem@4.1.3
 - gatsby-source-graphql@4.1.3
 - gatsby-source-shopify@6.1.4
 - gatsby-source-wordpress@6.1.3
 - gatsby-telemetry@3.1.3
 - gatsby-transformer-remark@5.1.4
 - gatsby-transformer-sqip@4.1.4
 - gatsby@4.1.6
 ```

**Reply by cameronmcefee** [reply]
Unfortunately, I'm still seeing the issue with `gatsby@4.1.6` and `gatsby-source-contentful@6.1.4` and a `gatsby clean`. I also tried nuking node_modules out of superstition.

Again the cache hash is the same, but the following number is off by a bit.

Requested: dcd26b940a6f037f531e3a870147ca84-5
On Disk: dcd26b940a6f037f531e3a870147ca84-2

```
/Users/cameronmcefee/github/static-sites/node_modules/yoga-layout-prebuilt/yoga-layout/build/Release/nbind.js:53
        throw ex;
        ^

Error: ENOENT: no such file or directory, open '/Users/cameronmcefee/github/static-sites/packages/blog/.cache/caches/gatsby-source-contentful/dcd26b940a6f037f531e3a870147ca84-5/Errors1-hero.jpg'
Emitted 'error' event on ReadStream instance at:
    at /Users/cameronmcefee/github/static-sites/node_modules/graceful-fs/graceful-fs.js:292:14
    at /Users/cameronmcefee/github/static-sites/node_modules/graceful-fs/graceful-fs.js:343:16
    at FSReqCallback.oncomplete (fs.js:180:23) {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/Users/cameronmcefee/github/static-sites/packages/blog/.cache/caches/gatsby-source-contentful/dcd26b940a6f037f531e3a870147ca84-5/Errors1-hero.jpg'

 ERROR #85928 

An error occurred during parallel query running.
Go here for troubleshooting tips: https://gatsby.dev/pqr-feedback
```

I'm still in the middle of migrating plugins to their Gatsby 4 supporting versions. Is it possible this could be something caused by an out of date plugin? I'm just assuming that at this point this is my fault some how 😄 .

**Reply by wardpeet** [reply]
Can you make sure you remove yarn.lock or package-lock.json. You might have multiple versions of gatsby-core-utils or perhaps it didn't update correctly. I couldn't reproduce the issue anymore

**Reply by cameronmcefee** [reply]
It's been a busy week so I haven't had a chance to do the work yet, but I can confirm I've got multiple versions of gatsby-core-utils from packages I haven't updated yet. I'll try to carve out some time to update everything this week so I can confirm whether things are good with that change.

**Reply by cameronmcefee** [reply]
This issue is resolved after updating a few other plugins so that I've only got one version of gatsby-core-utils. Thanks @wardpeet. And thank god for `yarn why` to figure out what they were.

**Reply by nonAlgebraic** [reply]
Hi! Just wanted to chime in and say that we're experiencing the same issue, but we have deduped `gatsby-core-utils` with a Yarn `resolution`, refreshed `yarn.lock`, deleted and reinstalled `node_modules`, and all of our plugins (and packages at large) are at their latest versions.

Might there be any clues to other reasons for this type of error?

**Reply by TylerBarnes** [reply]
@nonAlgebraic can you share the output of `yarn why gatsby-core-utils`? Also, which version did you set the resolution to?

**Reply by nonAlgebraic** [reply]
@TylerBarnes it's on 3.7.0.

```bash
[1/4] 🤔  Why do we have the module "gatsby-core-utils"...?
[2/4] 🚚  Initialising dependency graph...
[3/4] 🔍  Finding dependency...
[4/4] 🚡  Calculating file sizes...
=> Found "gatsby-core-utils@3.7.0"
info Has been hoisted to "gatsby-core-utils"
info Reasons this module exists
   - Specified in "dependencies"
   - Hoisted from "gatsby-transformer-inline-svg#gatsby-core-utils"
   - Hoisted from "gatsby-plugin-netlify#gatsby-core-utils"
   - Hoisted from "babel-preset-gatsby#gatsby-core-utils"
   - Hoisted from "gatsby#gatsby-core-utils"
   - Hoisted from "gatsby-cli#gatsby-core-utils"
   - Hoisted from "gatsby-plugin-image#gatsby-core-utils"
   - Hoisted from "gatsby-plugin-manifest#gatsby-core-utils"
   - Hoisted from "gatsby-plugin-sharp#gatsby-core-utils"
   - Hoisted from "gatsby-source-contentful#gatsby-core-utils"
   - Hoisted from "gatsby-source-filesystem#gatsby-core-utils"
   - Hoisted from "gatsby-source-shopify#gatsby-core-utils"
   - Hoisted from "gatsby-transformer-remark#gatsby-core-utils"
   - Hoisted from "gatsby#babel-plugin-remove-graphql-queries#gatsby-core-utils"
   - Hoisted from "gatsby#gatsby-page-utils#gatsby-core-utils"
   - Hoisted from "gatsby#gatsby-plugin-page-creator#gatsby-core-utils"
   - Hoisted from "gatsby#gatsby-telemetry#gatsby-core-utils"
info Disk size without dependencies: "292KB"
info Disk size with unique dependencies: "2.61MB"
info Disk size with transitive dependencies: "12.09MB"
info Number of shared dependencies: 62
```
